tests/support/client.rs

```diff
@@ -245,40 +271,135 @@ impl hyper::client::connect::Connect for Conn {
     type Error = ::std::io::Error;
     fn connect(&self, _: hyper::client::connect::Destination) -> Self::Future {
         let connected = hyper::client::connect::Connected::new().proxy(self.absolute_uris);
-        Box::new(self.connect_().map(|t| (t, connected)))
+        Box::new(self.initialize_connection().map(|t| (t, connected)))
     }
 }
 
-/// A wrapper around a TcpStream, allowing us to signal when the connection
-/// is dropped.
-struct RunningIo {
-    inner: TcpStream,
-    /// When this drops, the related Receiver is notified that the connection
-    /// is closed.
-    running: oneshot::Sender<()>,
+pub enum ConnectorFuture<F: Future> {
+    Init {
+        future: F,
+        tls: Option<(Arc<ClientConfig>, DNSName)>,
+        running: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+    },
+    Handshake {
+        future: tokio_rustls::Connect<F::Item>,
+        running: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+    },
+}
+
+impl<F> Future for ConnectorFuture<F>
+where
+    F: Future,
+    F::Item: AsyncRead + AsyncWrite + Send + 'static,
+    F::Error: From<io::Error>,
+{
+    type Item = RunningIo<F::Item>;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        fn take_running(r: &Mutex<Option<oneshot::Sender<()>>>) -> oneshot::Sender<()> {
+            r.lock()
+                .expect("running lock")
+                .take()
+                .expect("support client cannot connect more than once")
+        }
+
+        loop {
+            *self = match self {
+                ConnectorFuture::Init {
+                    future,
+                    tls,
+                    running,
+                } => {
+                    let io = try_ready!(future.poll());
+
+                    match tls {
+                        None => {
+                            return Ok(Async::Ready(RunningIo::Http(io, take_running(running))));
+                        }
+
+                        Some((cfg, name)) => {
+                            let future = tokio_rustls::TlsConnector::from(cfg.clone())
+                                .connect(DNSName::as_ref(name), io);
+                            ConnectorFuture::Handshake {
+                                future,
+                                running: running.clone(),
+                            }
+                        }
+                    }
+                }
+
+                ConnectorFuture::Handshake { future, running } => {
+                    let io = try_ready!(future.poll());
+                    return Ok(Async::Ready(RunningIo::Https(io, take_running(running))));
+                }
+            }
+        }
+    }
 }
 
-impl io::Read for RunningIo {
+pub enum RunningIo<T> {
```

Comment:
Again, not sure if this needs to be `pub`. 

Also, I think the underlying IO will always be a `TcpStream`, so I don't think it really needs to be generic over a `T`...